spring-kafka/src/main/java/org/springframework/kafka/listener/adapter/ShareRecordMessagingMessageListenerAdapter.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2002-present the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.kafka.listener.adapter;
+
+import java.lang.reflect.Method;
+
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.ShareConsumer;
+import org.jspecify.annotations.Nullable;
+
+import org.springframework.kafka.listener.AcknowledgingConsumerAwareMessageListener;
+import org.springframework.kafka.listener.AcknowledgingShareConsumerAwareMessageListener;
+import org.springframework.kafka.listener.KafkaListenerErrorHandler;
+import org.springframework.kafka.support.ShareAcknowledgment;
+import org.springframework.kafka.support.converter.JacksonProjectingMessageConverter;
+import org.springframework.kafka.support.converter.ProjectingMessageConverter;
+import org.springframework.kafka.support.converter.RecordMessageConverter;
+import org.springframework.messaging.Message;
+
+/**
+ * A {@link org.springframework.kafka.listener.MessageListener MessageListener}
+ * adapter that invokes a configurable {@link HandlerAdapter}; used when the factory is
+ * configured for the listener to receive individual messages from share groups.
+ *
+ * <p>Wraps the incoming Kafka Message to Spring's {@link Message} abstraction.
+ *
+ * <p>The original {@link ConsumerRecord} and
+ * the {@link ShareAcknowledgment} are provided as additional arguments so that these can
+ * be injected as method arguments if necessary.
+ *
+ * @param <K> the key type.
+ * @param <V> the value type.
+ * @author Soby Chacko
+ * @since 4.0
+ */
+public class ShareRecordMessagingMessageListenerAdapter<K, V> extends MessagingMessageListenerAdapter<K, V>
```

Comment:
If we can make ` @Nullable Object acknowledgment, @Nullable Object consumer` in the `MessagingMessageListenerAdapter`, then we might not need this class at all.
And might be not need even `AcknowledgingShareConsumerAwareMessageListener`.
I don't see how is that important to such a contract fine-grained for expected types.

Reply:
I see your point. However, I think, making this type of change might be too risky from a regression standpoint, as we have to change the API in `GenericMessageListener` to accept `Object` types. Don't you think that we are compromising on the type safety that the frameworks provides today? Plus, `MessagingMessageListenerAdapter` is an abstract class and we need a concrete adapter to deal with share consumers (as we have record and batch adapters). In any case, I think, we may have to look at this particular change after this PR is merged, but let's chat next week.

Reply:
OK. We can live with that, but `MessagingMessageConverter` is not very justified.